jena-fuseki2/jena-fuseki-core/src/main/java/org/apache/jena/fuseki/server/Dispatcher.java

```diff
@@ -203,151 +190,91 @@ private static ActionProcessor chooseProcessor(HttpAction action) {
-        // Try to find "R" functionality from an RW.
-        if ( GSP_R.equals(operation) )
-            return findEndpointForOperationExact(dataService, GSP_RW, preferUnnamed);
-        // Instead of 404, return 405 if asked for RW but only R available.
-        if ( GSP_RW.equals(operation) && dataService.hasOperation(GSP_R) )
-            ServletOps.errorMethodNotAllowed(action.getMethod());
-        return null;
-    }
-
-    /** Find a matching endpoint for exactly this operation.
-     * If multiple choices, prefer either named or unnamed according
-     * to the flag {@code preferUnnamed}.
-     */
-    private static Endpoint findEndpointForOperationExact(DataService dataService, Operation operation, boolean preferUnnamed) {
-        List<Endpoint> eps = dataService.getEndpoints(operation);
-        if ( eps == null || eps.isEmpty() )
-            return null;
-        // ==== Legacy compatibility.
-        // Find a named service, with preference for named/unnamed.
-        Endpoint epAlt = null;
-        for ( Endpoint ep : eps ) {
-            if ( operation.equals(ep.getOperation()) ) {
-                if ( ep.isUnnamed() && preferUnnamed )
-                    return ep;
-                if ( ! ep.isUnnamed() && ! preferUnnamed )
-                    return ep;
-                epAlt = ep;
+        if ( ep == null ) {
+            if ( GSP_R.equals(operation) )
+                // If asking for GSP_R, and GSP_RW available, pass that back.
+                ep = epSet.get(GSP_RW); // [GSP Promote]
+            else if ( GSP_RW.equals(operation) ) {
+                // If asking for GSP_RW, only GSP_R available -> 405.
+                if ( epSet.contains(GSP_R) )
+                    ServletOps.errorMethodNotAllowed(action.getMethod());
             }
         }
-        // Did not find a preferred one.
-        return epAlt;
+
+        // There are multiple endpoints; if none are suitable, then 400.
+        if ( ep == null )
+            ServletOps.errorBadRequest("No operation for request: "+action.getActionURI());
+        return ep;
     }
 
     /**
      * Identify the operation being requested.
      * It is analysing the HTTP request using global configuration.
      * The decision is based on
      * <ul>
-     * <li>Query parameters (URL query string or HTML form)</li>
-     * <li>Content-Type header</li>
-     * <li>Otherwise it is a plain REST (quads) operation.chooseOperation</li>
+     * <li>HTTP query string parameters (URL query string or HTML form)</li>
+     * <li>Registered Content-Type header</li>
+     * <li>Otherwise it is a plain REST (quads)</li>
      * </ul>
      * The HTTP Method is not considered.
      * <p>
      * The operation is not guaranteed to be supported on every {@link DataService}
      * nor that access control will allow it to be performed.
      */
-    public static Operation chooseOperation(HttpAction action) {
+    private static Operation chooseOperation(HttpAction action) {
```

Comment:
Any particular reason to lock things down to `private` ?

Reply:
Only that this is an internal operation. From what I can tell, it is a left over from when dispatch was split across multiple classes.